manifest/list.go

```diff
@@ -0,0 +1,119 @@
+// directly.
+type ManifestList interface {
+	// MIMEType returns the MIME type of this particular manifest list.
+	MIMEType() string
+
+	// Instances returns a list of the manifests that this list knows of.
+	Instances() []types.BlobInfo
+
+	// Update information about the list's instances.  The length of the passed-in slice must
+	// match the length of the list of instances which the list already contains, and every field
+	// must be specified.
+	UpdateInstances([]ManifestListUpdate) error
+
+	// ChooseDigest selects which manifest is most appropriate for the platform described by the
+	// SystemContext, or for the current platform if the SystemContext doesn't specify any details.
+	ChooseDigest(ctx *types.SystemContext) (digest.Digest, error)
+
+	// ImageID computes a recommended image ID based on the list of images referred to by the manifest.
+	ImageID() string
+
+	// Serialize returns the list in a blob format.
+	// NOTE: Serialize() does not in general reproduce the original blob if this object was loaded
+	// from, even if no modifications were made!
+	Serialize() ([]byte, error)
+
+	// ToOCI1Index() returns the list rebuilt as an OCI1 index, converting it if necessary.
+	ToOCI1Index() (*OCI1Index, error)
+
+	// ToSchema2List() returns the list rebuilt as a Schema2 list, converting it if necessary.
+	ToSchema2List() (*Schema2List, error)
+}
+
+// ManifestListUpdate includes the fields which a manifest's UpdateInstances() method will modify.
+type ManifestListUpdate struct {
+	Digest    digest.Digest
+	Size      int64
+	MediaType string
+}
+
+// dupStringSlice returns a deep copy of a slice of strings, or nil if the
+// source slice is empty.
+func dupStringSlice(list []string) []string {
+	if len(list) == 0 {
+		return nil
+	}
+	dup := make([]string, len(list))
+	for i := range list {
+		dup[i] = list[i]
+	}
+	return dup
+}
+
+// dupStringStringMape returns a deep copy of a map[string]string, always returning
+// an initialized map, even if there are no contents to copy into it.
+func dupStringStringMap(m map[string]string) map[string]string {
+	if len(m) == 0 {
+		return nil
+	}
+	result := make(map[string]string)
+	for k, v := range m {
+		result[k] = v
+	}
+	return result
+}
+
+// ListFromBlob parses a list of manifests.
+func ListFromBlob(manifest []byte, manifestMIMEType string) (ManifestList, error) {
+	switch normalized := NormalizedMIMEType(manifestMIMEType); normalized {
+	case DockerV2ListMediaType:
+		return Schema2ListFromManifest(manifest)
+	case imgspecv1.MediaTypeImageIndex:
+		return OCI1IndexFromManifest(manifest)
+	case DockerV2Schema1MediaType, DockerV2Schema1SignedMediaType:
+		return nil, fmt.Errorf("Treating single images as manifest lists is not implemented")
+	case imgspecv1.MediaTypeImageManifest:
+		return nil, fmt.Errorf("Treating single images as manifest lists is not implemented")
+	case DockerV2Schema2MediaType:
+		return nil, fmt.Errorf("Treating single images as manifest lists is not implemented")
```

Comment:
There isn't AFAIK.  This was mimicking the behavior in `FromBlob()`, but since we're not customizing the error message based on the type, you're right, there's no need for that.  Collapsed them.